[Relay][Frontend][Onnx] GRU Layer Support

jwfromm opened the pull request:
This PR adds GRU parsing to the onnx frontend. Currently this is done by unrolling the recurrence in a similar way to how we handle LSTMs. Since there's quite a bit of shared code, I've generalized the LSTM converter to an RNN converter and made LSTM and GRU subclasses. For testing, I again replaced the LSTM test function with a more general test_rnn function that supports both LSTM and GRU and should be easily expandable to other RNN functions should we ever want to add any.

jwfromm commented:
@masahi, @areusch can you guys take a look at this PR?

anijain2305 commented:
This might be somewhat relevant - https://discuss.tvm.ai/t/onnx-lstm-op-conversion/7238
Just pointing it out here. I think the GRU might also have the same problems that I encounter with LSTM.

masahi commented:
@siju-samuel please help review this PR.

jwfromm commented:
I've converted this to a draft as I'll try to incorporate the points made by @anijain2305 in the linked discuss post.

areusch commented on python/tvm/relay/frontend/onnx.py:
```diff
@@ -1530,6 +1532,14 @@ def _activation_needs_beta(cls, activation):
 
     @classmethod
     def _impl_v7(cls, inputs, attr, params):
+        if cls.name == 'LSTM':
```
here I think it'd be cleaner to just implement the `@classmethod` in each subclass, no?

jwfromm replied:
you're right thats a nicer organization. I've updated the structure accordingly.

areusch commented on python/tvm/relay/frontend/onnx.py:
```diff
@@ -1530,6 +1532,14 @@ def _activation_needs_beta(cls, activation):
+        raise NotImplementedError("%s RNNs are not yet supported." % cls.name)
+
+    @classmethod
+    def _lstm(cls, inputs, attr, params):
```
any reason not to move this to LSTM subclass?

jwfromm commented:
I've added a small change to how arguments are parsed in RNNs based on the discussion here: https://discuss.tvm.ai/t/onnx-lstm-op-conversion/7238/5. Our previous implementation assumed that the position of optional arguments could not be known without their name, however, this is not true assuming the nodes are constructing properly. To avoid name based problems, RNNs now use indexing to get inputs. I've updated the `onnx_input` structure to allow indexing outside of the input bounds but return `None` in such cases, indicating that an input was not provided.

masahi commented:
Thanks @jwfromm @areusch @anijain2305
